test/integration/fake_upstream.cc

```diff
@@ -413,6 +414,43 @@ void FakeHttpConnection::onGoAway(Http::GoAwayErrorCode code) {
   ENVOY_LOG(info, "FakeHttpConnection receives GOAWAY: ", code);
 }
 
+void FakeHttpConnection::encodeGoAway() {
+  ASSERT(type_ == Type::HTTP2);
+
+  shared_connection_.connection().dispatcher().post([this]() { codec_->goAway(); });
+}
+
+void FakeHttpConnection::encodeProtocolError() {
+  ASSERT(type_ == Type::HTTP2);
+
+  Http::Http2::ServerConnectionImpl* codec =
+      dynamic_cast<Http::Http2::ServerConnectionImpl*>(codec_.get());
+  if (codec != nullptr) {
+    shared_connection_.connection().dispatcher().post([this, codec]() {
+      expect_protocol_error_ = true;
```

Comment:
@mattklein123 this is basically the same approach as before, but I was able to move the hack into the fake upstream.

What the issue was is that when the nghttp2 callback fails, the upstream end of the connection gets closed, which then creates a race from the perspective of the health checker: it will either see the GOAWAY first and close its end of thee connection (good) or see the reset from the upstream (bad, causes timing issues in the test).

It's still not ideal, but it's at least limited only to the fake upstream, and this flag is only set in encodeProtocolError(). The benefit is that other tests that don't need to exercise protocol errors won't have their behavior changed.

Also I just wanted to say thanks for all of the time you've spent reviewing, I appreciate it. I've learned a lot from this PR so hopefully my next contribution will require less maintainer time :)

Reply:
Sorry which callback fails? You mean when you send the GOAWAY, this causes the fake upstream to close, and that sometimes races with the GOAWAY being received by the health checker? I guess we are doing a close without flush in this case? How often does it flake when you run the test with runs_per_test?

Reply:
The `onFrameSend()` callback (which is what I modified in my previous approach) is what fails. Then in `onData()` in FakeHttpConnection::ReadFilter it closes the connection which causes the race. It failed ~1-3 times with --runs_per_test=500

I don't think the issue is that the frame not getting flushed, I'd consistently see "connection going away" in the logs on a failed run. It was just a timing issue where in the integration test we have this sequence:

```
  ASSERT_TRUE(clusters_[cluster_idx].host_fake_connection_->waitForDisconnect());
  timeSystem().advanceTimeWait(std::chrono::milliseconds(500));
```

We want `waitForDisconnect()` to mean that the health checker has seen the GOAWAY frame, and then enabled the interval timer to set up the next check. However on a test flake run, waitForDisconnect() gets triggered by the fake upstream itself closing, so then the clock is advanced before the interval timer has been enabled. If there is a different way of waiting for the health checker to see the frame and enable thee timer then that would also work, although that kind of takes some of the value out of asserting a disconnect, we do want to make sure that the health checker closes the connection immediately.

Reply:
I see. Can you just wait on a stat emitted by the health checker? Then wait on the disconnect?

Reply:
I would need to reorder when stats are incremented, because that happens before the interval timer is enabled. Would that be preferable to this change?

Reply:
I think it would be less hacky, but why does the stat need to be reordered? Is the issue that you need the connection attempted stat because this doesn't count as a failure?

Reply:
the issue is that we're using simulated time, so I have to call `  timeSystem().advanceTimeWait(std::chrono::milliseconds(500));` to advance time to trigger another health check. But that only works if the timer has been enabled before I advance the time system. So I need something I can wait on in the test that tells me that the health checker has enabled the timer. The client connection close does happen after the timer is enabled, but stats are incremented before which doesn't make them suitable to wait on.

Reply:
I'm not following you. What stats are available? Is the issue that there is no stat when the GOAWAY is received?

Reply:
Basically the existing stats are incremented too early for them to be useful for my test to wait on, it may still be "too soon" to advance time.

If I add a new stat for GOAWAY errors (which is maybe useful anyway?) I can increment it after the health check interval timer is enabled and solve it that way. I'll push a new commit with that approach

Reply:
Yes I would add a stat for "close_notify" or similar like we do on the data plane side. Please check that stat name. Also, please remember to doc the new stat. Otherwise sgtm. Thanks!

Reply:
I didn't see this comment until after my most recent commit with the stat named `goaway_error`. It looks like `close_notify` is used for graceful close, e.g. `Connection: close` in HTTP/1. Here we're talking about an ungraceful close since the GOAWAY has a code other than `NO_ERROR`. Is that stat name still appropriate? If so, I should probably also increment it in the NO_ERROR and HTTP/1 Connection: close cases

Reply:
I don't see where you incremented the new stat anywhere in the code you pushed. Also, you have lost me at this point. If this is a sub-class of a network error why can't you wait on the failure stat?

Reply:
I apparently did not commit the 1 line where I incremented it 🤦

Reply:
I can use the failure stat instead, IF I move the the stat `inc()` call to after `interval_timer_->enableTimer();`